vta/hardware/chisel/src/main/scala/core/TensorAlu.scala

```diff
@@ -0,0 +1,316 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package vta.core
+
+import chisel3._
+import chisel3.util._
+import vta.util.config._
+
+class Alu(implicit p: Parameters) extends Module {
+  val aluBits = p(CoreKey).accBits
+  val io = IO(new Bundle {
+    val opcode = Input(UInt(C_ALU_OP_BITS.W))
+    val a = Input(SInt(aluBits.W))
+    val b = Input(SInt(aluBits.W))
+    val y = Output(SInt(aluBits.W))
+  })
+  val min = Mux(io.a < io.b, io.a, io.b)
+  val max = Mux(io.a < io.b, io.b, io.a)
+  val add = io.a + io.b
+
+  val ub = io.b.asUInt
+  val width = log2Ceil(aluBits)
+  val m = ~ub(width - 1, 0) + 1.U // FIXME: there should be a SHL and SHR instruction
+  val n = ub(width - 1, 0)
+  val shl = io.a << m
+  val shr = io.a >> n
+
+  io.y :=
+    MuxLookup(io.opcode,
+              io.a, // default
+      Array(
+        ALU_OP_MIN -> min,
+        ALU_OP_MAX -> max,
+        ALU_OP_ADD -> add,
+        ALU_OP_SHR -> shr,
+	ALU_OP_SHL -> shl
```

Comment:
This may not compatible with the HLS implement. What is the pros and cons if we implement this with a sign checker of `ub(width - 1, 0)`?

Reply:
The "cleanest way" I found for implementing that "negative shift" was [here](https://github.com/vegaluisjose/tvm/blob/nVTA-pr/vta/hardware/chisel/src/main/scala/core/TensorAlu.scala#L224-L226)

The real solution to this problem is that the ISA should support both SHR and SHL instruction in the first place and not rely on the sign of the operand. However, adding that right now will require other changes that might complicate more the PR. I have this in my TODO list for future improvements.